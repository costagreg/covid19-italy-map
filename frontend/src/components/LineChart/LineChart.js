import React, { Component, Fragment, useRef, useEffect } from 'react'
import classnames from 'classnames'
import PropTypes from 'prop-types'

import { scaleLinear } from 'd3-scale'
import { max, min } from 'd3-array'
import { line, curveCatmullRom } from 'd3-shape'
import { format } from 'date-fns'

import './LineChart.scss'

export const ToolTip = ({
  xPos,
  yPos,
  xValue,
  yValue,
  label,
  chartWidth,
  chartHeight,
  toolTipWidth,
  toolTipHeight,
}) => {
  xPos =
    xPos + toolTipWidth < chartWidth
      ? xPos + 10
      : chartWidth - (toolTipWidth + 20)
  yPos =
    yPos + toolTipHeight < chartHeight
      ? yPos + 10
      : chartHeight - (toolTipHeight + 20)

  return (
    <foreignObject
      data-testid="toolTip"
      x={xPos}
      y={yPos}
      width="200"
      height={100}
    >
      <div className="toolTip">
        <div className="toolTip__yValue">
          <b className="toolTip__yValueLabel">{label}</b>:
          <span className="toolTip__yValueText">{yValue}</span>
        </div>
        <div className="toolTip__xValue">{xValue}</div>
      </div>
    </foreignObject>
  )
}

class LineChart extends Component {
  state = {
    pointHovered: -1,
  }
  render() {
    const { pointHovered } = this.state
    const { xAxis, yAxis, chartWidth, chartHeight, margin, yLabel } = this.props

    const [marginTop, marginRight, marginBottom, marginLeft] = margin

    const xAxisScale = scaleLinear()
      .domain([0, xAxis.length - 1])
      .range([marginLeft, chartWidth - marginRight])

    const yAxisScale = scaleLinear()
      .domain([Math.min(min(yAxis), 0), max(yAxis)])
      .range([chartHeight - marginBottom, marginTop])

    const chartLine = line()
      .x((d, i) => xAxisScale(i))
      .y((d) => yAxisScale(d))
      .curve(curveCatmullRom.alpha(0.5))

    // console.log(yAxis)

    const xTicks = xAxisScale.ticks(6).map((d) =>
      xAxis[d] ? (
        <g
          key={xAxis[d]}
          transform={`translate(${xAxisScale(d)},${chartHeight -
            marginBottom})`}
        >
          <text className="lineChart__xTicksText">
            {format(new Date(+xAxis[d]), 'dd/MM')}
          </text>
          <line
            className="lineChart__xTicksLine"
            x1="0"
            x1="0"
            y1="0"
            y2="5"
          />
        </g>
      ) : null
    )

    return (
      <div className="lineChart">
        <svg
          width={chartWidth}
          height={chartHeight}
          viewBox={`0 0 ${chartWidth} ${chartHeight}`}
        >
          <g>
            <line
              className="lineChart__xAxis"
              x1={marginLeft}
              x2={chartWidth - marginRight}
              y1={chartHeight - marginBottom}
              y2={chartHeight - marginBottom}
            />
          </g>
          <g>
            <path className="lineChart__line" d={chartLine(yAxis)} />
            {yAxis.map((d, i) => (
              <g
                key={xAxis[i]}
                transform={`translate(${xAxisScale(i)},${yAxisScale(d)})`}
              >
                <circle
                  className={classnames([
                    'lineChart__circle',
                    { lineChart__circleSelected: i === pointHovered },
                  ])}
                  r={2}
                />
                <circle
                  data-testid="dataPoint"
                  r={10}
                  fill="transparent"
                  onMouseOver={() => {
                    this.setState({ pointHovered: i })
                  }}
                  onMouseOut={() => {
                    this.setState({ pointHovered: -1 })
                  }}
                />
              </g>
            ))}
          </g>
          {xTicks}
          {pointHovered >= 0 ? (
            <ToolTip
              xPos={xAxisScale(pointHovered)}
              yPos={yAxisScale(yAxis[pointHovered])}
              xValue={format(new Date(+xAxis[pointHovered]), 'dd/MM/yy')}
              yValue={yAxis[pointHovered]}
              label={yLabel}
              chartWidth={chartWidth}
              chartHeight={chartHeight}
              toolTipWidth={200}
              toolTipHeight={100}
            />
          ) : null}
        </svg>
      </div>
    )
  }
}

LineChart.protoTypes = {
  xAxis: PropTypes.arrayOf(PropTypes.number).isRequired,
  yAxis: PropTypes.arrayOf(
    PropTypes.oneOf([PropTypes.string, PropTypes.number])
  ).isRequired,
  chartWidth: PropTypes.number.isRequired,
  chartHeight: PropTypes.number.isRequired,
  margin: PropTypes.arrayOf(PropTypes.number),
  yLabel: PropTypes.arrayOf(PropTypes.string),
}

LineChart.defaultProps = {
  margin: [0, 0, 0, 0],
  yLabel: '',
}

export default LineChart