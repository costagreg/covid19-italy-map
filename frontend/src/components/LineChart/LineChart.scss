.lineChart{
  &__xAxis{
    stroke:#000;
  }
  &__line{
    stroke: steelblue;
    stroke-width: 2px;
    fill: none;
  }

  &__circleSelected{
    fill: red;
    r:4;
  }

  &__xTicksText{
    font-size: 12px;
    transform:translate(0,15px)
  }

  &__xTicksLine{
    stroke:#000;
    transform:translate(0,0)
  }

  foreignObject{
    overflow:visible;
  }

  .toolTip{
    text-align: center;
    opacity: 0.9;
    font-size: 0.7rem;
    box-shadow: 0px 0px 9px 0px rgba(24, 9, 9, 0.2);
    background: #fff;
    padding: 10px;
    &__xValue{
      display: block;
    }
    &__yValue{    
      // display: flex;
      flex-direction: row;
    }
  }

}