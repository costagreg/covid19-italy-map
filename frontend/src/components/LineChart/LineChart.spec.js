import { render, fireEvent } from '@testing-library/react'
import React from 'react'
import LineChart, { ToolTip } from './LineChart'

jest.mock('../../constants', () => ({
  params: { labelMock: 'label' },
}))

describe('LineChart', () => {
  const initProps = {
    xAxis: [1586448000000, 1586534400000, 1586620800000],
    yAxis: [100, 120, 110],
    chartWidth: 300,
    chartHeight: 200,
    margin: [0, 0, 0, 0],
    yLabel: 'labelMock',
  }

  describe('@render', () => {
    it('renders LineChart without any error', () => {
      const { asFragment } = render(<LineChart {...initProps} />)

      expect(asFragment()).toMatchSnapshot()
    })
  })

  describe('@events', () => {
    it('shows the tooltip when user hover mouse over to a data point', () => {
      const { getAllByTestId, debug, queryByText } = render(
        <LineChart {...initProps} />
      )

      const datePoints = getAllByTestId('dataPoint')

      expect(queryByText('label')).toBeNull()
      expect(queryByText(`${initProps.yAxis[0]}`)).toBeNull()

      fireEvent.mouseOver(datePoints[0])

      expect(queryByText('label')).toBeTruthy()
      expect(queryByText(`${initProps.yAxis[0]}`)).toBeTruthy()
      expect(queryByText(`${initProps.yAxis[1]}`)).toBeNull()
      expect(queryByText(`${initProps.yAxis[2]}`)).toBeNull()
    })

    it('hide the tooltip when user hover mouse out to a data point', () => {
      const { getAllByTestId, queryByText } = render(
        <LineChart {...initProps} />
      )

      const datePoints = getAllByTestId('dataPoint')

      expect(queryByText('label')).toBeNull()
      expect(queryByText(`${initProps.yAxis[2]}`)).toBeNull()

      fireEvent.mouseOver(datePoints[2])

      expect(queryByText('label')).toBeTruthy()
      expect(queryByText(`${initProps.yAxis[0]}`)).toBeNull()
      expect(queryByText(`${initProps.yAxis[1]}`)).toBeNull()
      expect(queryByText(`${initProps.yAxis[2]}`)).toBeTruthy()

      fireEvent.mouseOut(datePoints[2])

      expect(queryByText('label')).toBeNull()
      expect(queryByText(`${initProps.yAxis[2]}`)).toBeNull()
    })
  })
})

describe('ToolTip', () => {
  const isToolTipNotCropped = (toolTip, { chartWidth, chartHeight }) => {
    const xPosToolTip = parseInt(toolTip.getAttribute('x'))
    const yPosToolTip = parseInt(toolTip.getAttribute('y'))

    expect(xPosToolTip).toBeLessThanOrEqual(chartWidth)
    expect(yPosToolTip).toBeLessThanOrEqual(chartHeight)
    expect(xPosToolTip).toBeGreaterThanOrEqual(0)
    expect(yPosToolTip).toBeGreaterThanOrEqual(0)
  }

  describe('@render', () => {
    const initProps = {
      xPos: 10,
      yPos: 20,
      xValue: '12/06/1992',
      yValue: 60,
      label: 'test',
      chartWidth: 300,
      chartHeight: 200,
      toolTipWidth: 50,
      toolTipHeight: 50,
    }

    it('renders the ToolTip without any error', () => {
      const { asFragment } = render(<ToolTip {...initProps} />)

      expect(asFragment()).toMatchSnapshot()
    })

    it('does not crop the tooltip if datapoint is close to left top edge', () => {
      const props = {
        ...initProps,
        xPos: 0,
        yPos: 0,
      }

      const { getByTestId } = render(<ToolTip {...props} />)

      isToolTipNotCropped(getByTestId('toolTip'), props)
    })

    it('does not crop the tooltip if datapoint is close to right top edge', () => {
      const props = {
        ...initProps,
        xPos: 290,
        yPos: 10,
      }

      const { getByTestId } = render(<ToolTip {...props} />)

      isToolTipNotCropped(getByTestId('toolTip'), props)
    })

    it('does not crop the tooltip if datapoint is close to right bottom edge', () => {
      const props = {
        ...initProps,
        xPos: 290,
        yPos: 190,
      }

      const { getByTestId } = render(<ToolTip {...props} />)

      isToolTipNotCropped(getByTestId('toolTip'), props)
    })

    it('does not crop the tooltip if datapoint is close to left bottom edge', () => {
      const props = {
        ...initProps,
        xPos: 10,
        yPos: 190,
      }

      const { getByTestId } = render(<ToolTip {...props} />)

      isToolTipNotCropped(getByTestId('toolTip'), props)
    })
  })
})
