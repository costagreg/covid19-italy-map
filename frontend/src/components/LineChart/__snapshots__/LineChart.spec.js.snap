// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LineChart @render renders LineChart without any error 1`] = `
<DocumentFragment>
  <div
    class="lineChart"
  >
    <svg
      height="200"
      viewBox="0 0 300 200"
      width="300"
    >
      <g>
        <line
          class="lineChart__xAxis"
          x1="0"
          x2="300"
          y1="200"
          y2="200"
        />
      </g>
      <g>
        <path
          class="lineChart__line"
          d="M0,33.33333333333333C0,33.33333333333333,99.77340001969233,2.7151710776023803,150,0C199.7774347633459,-2.690889911713435,300,16.666666666666675,300,16.666666666666675"
        />
        <g
          transform="translate(0,33.33333333333333)"
        >
          <circle
            class="lineChart__circle"
            r="2"
          />
          <circle
            data-testid="dataPoint"
            fill="transparent"
            r="10"
          />
        </g>
        <g
          transform="translate(150,0)"
        >
          <circle
            class="lineChart__circle"
            r="2"
          />
          <circle
            data-testid="dataPoint"
            fill="transparent"
            r="10"
          />
        </g>
        <g
          transform="translate(300,16.666666666666675)"
        >
          <circle
            class="lineChart__circle"
            r="2"
          />
          <circle
            data-testid="dataPoint"
            fill="transparent"
            r="10"
          />
        </g>
      </g>
      <g
        transform="translate(0,200)"
      >
        <text
          class="lineChart__xTicksText"
        >
          09/04
        </text>
        <line
          class="lineChart__xTicksLine"
          x1="0"
          y1="0"
          y2="5"
        />
      </g>
      <g
        transform="translate(150,200)"
      >
        <text
          class="lineChart__xTicksText"
        >
          10/04
        </text>
        <line
          class="lineChart__xTicksLine"
          x1="0"
          y1="0"
          y2="5"
        />
      </g>
      <g
        transform="translate(300,200)"
      >
        <text
          class="lineChart__xTicksText"
        >
          11/04
        </text>
        <line
          class="lineChart__xTicksLine"
          x1="0"
          y1="0"
          y2="5"
        />
      </g>
    </svg>
  </div>
</DocumentFragment>
`;

exports[`ToolTip @render renders the ToolTip without any error 1`] = `
<DocumentFragment>
  <foreignobject
    data-testid="toolTip"
    height="100"
    width="200"
    x="20"
    y="30"
  >
    <div
      class="toolTip"
    >
      <div
        class="toolTip__yValue"
      >
        <b
          class="toolTip__yValueLabel"
        >
          test
        </b>
        :
        <span
          class="toolTip__yValueText"
        >
          60
        </span>
      </div>
      <div
        class="toolTip__xValue"
      >
        12/06/1992
      </div>
    </div>
  </foreignobject>
</DocumentFragment>
`;
